### PR TITLE
Adds support for leveraging StarTree index in conjunction with filtered aggregations

### PR DESCRIPTION
**Pull request by egalpin**
This PR aims to add support for leveraging StarTree index from filtered aggregations (i.e. `SELECT COUNT(*) FILTER (WHERE foo='bar')`), including filtered group-by aggregations.  For queries with a mixture of star-tree-solvable aggregations and aggs which cannot be solved by star-tree, those aggs solvable via star-tree will do so _iff_ all aggs in the same filter swimlane are solvable via star-tree (where swimlane means same `FILTER WHERE` expression, including lack of `FILTER WHERE`). 

tags: [`feature`]

### CODE REVIEW

File: pinot-core/src/main/java/org/apache/pinot/core/plan/GroupByPlanNode.java [hunk 1]
```diff
@@ -75,35 +70,10 @@ private GroupByOperator buildNonFilteredGroupByPlan() {
 
     FilterPlanNode filterPlanNode = new FilterPlanNode(_indexSegment, _queryContext);
     BaseFilterOperator filterOperator = filterPlanNode.run();
-
-    // Use star-tree to solve the query if possible
-    List<StarTreeV2> starTrees = _indexSegment.getStarTrees();
-    if (starTrees != null && !_queryContext.isSkipStarTree()) {
```
Comment (egalpin) [inline pinot-core/src/main/java/org/apache/pinot/core/plan/GroupByPlanNode.java:81]: Note: this PR would introduce a behaviour change (bugfix?) where `AggregationPlanNode` included an additional predicate to check if null handling is enabled before allowing use of StarTree index to solve the query.  This PR consolidates this logic to one method within OperatorUtils.java, where the null handling predicate is kept such that both GroupByPlanNode and AggregationPlanNode will ensure null handling is disabled before allowing use of StarTree index.

> if (starTrees != null && !_queryContext.isSkipStarTree() && !_queryContext.isNullHandlingEnabled()) {

**Comment by egalpin** [comment]
Marking as ready-for-review to get test results, but this should effectively be considered a draft at this point

**Comment by xiangfu0** [comment]
not sure if this is related: https://github.com/apache/pinot/issues/11899

**Comment by egalpin** [comment]
@xiangfu0 yep! The goal of this PR would be to support exactly what’s described in https://github.com/apache/pinot/issues/11899. Happy to collaborate, just let me know!

**Comment by xiangfu0** [comment]
> @xiangfu0 yep! The goal of this PR would be to support exactly what’s described in #11899. Happy to collaborate, just let me know!

Great!

File: pinot-core/src/main/java/org/apache/pinot/core/operator/query/OperatorUtils.java [hunk 2]
```diff
@@ -0,0 +1,124 @@
...
+    }
+
+    Set<ExpressionContext> expressionsToTransform =
+        AggregationFunctionUtils.collectExpressionsToTransform(aggregationFunctions, groupByExpressionsList);
+    return new ProjectPlanNode(indexSegment, queryContext, expressionsToTransform, DocIdSetPlanNode.MAX_DOC_PER_CALL,
+        filterOperator).run();
+  }
+
+  public static BaseProjectOperator<?> getProjectionOperator(
+      QueryContext queryContext,
```
Comment (egalpin) [inline pinot-core/src/main/java/org/apache/pinot/core/operator/query/OperatorUtils.java:88]: Note to self: pass in filter itself, not queryContext

**Comment by egalpin** [comment]
@xiangfu0 @Jackie-Jiang This can now be considered ready for review, as it's operationally functional.  I'll add tests, but wanted to flag that this is ready for review in case that helps with velocity etc.  See screen shots below for examples.

I noticed that this feature also provides an additional enhancement: being able to abuse `FILTER` aggs to allow queries to solve any agg which might be solvable via ST index even if one of the aggs in the query must be solved without the ST.  This isn't currently possible; currently, all aggs must be solvable by the ST in order to proceed with its use.  

Take for example the airlineStats example data.  The following query would utilize ST:

```SQL
select
  AirlineID,
  MAX(ArrDelay) as max_delay
FROM airlineStats
GROUP BY AirlineID
```

However, this would not (because `MIN` is not in the ST):

```SQL
select
  AirlineID,
  MAX(ArrDelay) as max_delay,
  MIN(ArrDelay)
FROM airlineStats
GROUP BY AirlineID
```

After this change, one could rewrite the query like so:

```SQL
select
  AirlineID,
  MAX(ArrDelay) FILTER (WHERE AirlineID != -1) AS max_delay,
  MIN(ArrDelay)
FROM airlineStats
GROUP BY AirlineID
```

This would allow the `MAX` to leverage the ST.

**Comment by egalpin** [comment]
Contrived example to show the working of FILTER (only has a result for the one row that matches `AirlineID`)

Screenshot 2023-11-06 at 16 26 40
Screenshot 2023-11-06 at 16 27 18

**Comment by egalpin** [comment]
@Jackie-Jiang @xiangfu0 this is now complete with tests and ready for review.  Thanks!

**Review by Jackie-Jiang — COMMENTED** [review]
Nice to see it fits to the current filtered aggregation architecture!

File: pinot-core/src/main/java/org/apache/pinot/core/operator/query/OperatorUtils.java [hunk 3]
```diff
@@ -0,0 +1,89 @@
...
+import org.apache.pinot.segment.spi.index.startree.StarTreeV2;
+
+
+public class OperatorUtils {
+  //TODO(egalpin): This should be merged with org.apache.pinot.core.operator.ProjectionOperatorUtils
+  private OperatorUtils() {
+    // Prevent instantiation, make checkstyle happy
+  }
+
+  public static ProjectionPlanNode maybeGetStartreeProjectionOperator(
```
[the remark this answers is not in the document]
Comment (egalpin) [reply pinot-core/src/main/java/org/apache/pinot/core/operator/query/OperatorUtils.java:51]: In that case would you put the onus of creating the `ProjectPlanNode` on the caller of `createStarTreeBasedProjectOperator` if it returned null?

File: pinot-integration-tests/src/test/java/org/apache/pinot/integration/tests/startree/StarTreeQueryGenerator.java [hunk 4]
```diff
@@ -82,6 +83,16 @@ private String generateAggregation(String metricColumn) {
         metricColumn);
   }
 
+  private String generateFilteredAggregation(String metricColumn) {
```
Comment (egalpin) [inline pinot-integration-tests/src/test/java/org/apache/pinot/integration/tests/startree/StarTreeQueryGenerator.java:86]: @Jackie-Jiang these tests seem to consistently fail when run using the multi-stage engine.  Are filtered aggregations supported in multi-stage, or not yet?

Comment (egalpin) [reply pinot-integration-tests/src/test/java/org/apache/pinot/integration/tests/startree/StarTreeQueryGenerator.java:86]: attempted fix in https://github.com/apache/pinot/pull/11886/commits/990bf5a725ef50c256f228e9415eea49566870fe

File: pinot-core/src/main/java/org/apache/pinot/core/query/aggregation/function/AggregationFunctionUtils.java [hunk 5]
```diff
@@ -214,46 +214,69 @@ public static Object getConvertedFinalResult(DataTable dataTable, ColumnDataType
     }
   }
```
Comment (egalpin) [inline pinot-core/src/main/java/org/apache/pinot/core/query/aggregation/function/AggregationFunctionUtils.java:216]: Much better, I was not liking the nested Pair usage haha. Thanks!

**Review by Jackie-Jiang — APPROVED** [review]

**Pull request merged**